third_party/internal/godocfx/parse.go

```diff
@@ -0,0 +1,297 @@
+}
+
+// child represents an item child.
+type child string
+
+// syntax represents syntax.
+type syntax struct {
+	Content string
+}
+
+// item represents a DocFX item.
+type item struct {
+	UID      string   `yaml:"uid"`
+	Name     string   `yaml:"name,omitempty"`
+	ID       string   `yaml:"id,omitempty"`
+	Summary  string   `yaml:"summary,omitempty"`
+	Parent   string   `yaml:"parent,omitempty"`
+	Type     string   `yaml:"type,omitempty"`
+	Langs    []string `yaml:"langs,omitempty"`
+	Syntax   syntax   `yaml:"syntax,omitempty"`
+	Children []child  `yaml:"children,omitempty"`
+}
+
+func (p *page) addItem(i *item) {
+	p.Items = append(p.Items, i)
+}
+
+func (i *item) addChild(c child) {
+	i.Children = append(i.Children, c)
+}
+
+var onlyGo = []string{"go"}
+
+// parse parses the directory into a map of import path -> page and a TOC.
+//
+// glob is the path to parse, usually ending in `...`. glob is passed directly
+// to packages.Load as-is.
+func parse(glob string) (map[string]*page, tableOfContents, *packages.Module, error) {
+	config := &packages.Config{
+		Mode:  packages.NeedName | packages.NeedSyntax | packages.NeedTypes | packages.NeedTypesInfo | packages.NeedModule,
+		Tests: true,
+	}
+
+	pkgs, err := packages.Load(config, glob)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("packages.Load: %v", err)
+	}
+	packages.PrintErrors(pkgs) // Don't fail everything because of one package.
+
+	if len(pkgs) == 0 {
+		return nil, nil, nil, fmt.Errorf("pattern %q matched 0 packages", glob)
+	}
+
+	pages := map[string]*page{}
+	toc := tableOfContents{}
+
+	module := pkgs[0].Module
+	skippedModules := map[string]struct{}{}
+
+	log.Printf("Processing %s@%s", module.Path, module.Version)
+
+	for _, pkg := range pkgs {
+		if pkg == nil || pkg.Module == nil {
+			continue
+		}
+		if pkg.Module.Path != module.Path {
+			skippedModules[pkg.Module.Path] = struct{}{}
+			continue
+		}
+		// Don't generate docs for tests or internal.
+		switch {
+		case strings.HasSuffix(pkg.ID, ".test"),
+			strings.HasSuffix(pkg.ID, ".test]"),
+			strings.Contains(pkg.ID, "internal"):
+			continue
+		}
+
+		// Collect all .go files.
+		files := []*ast.File{}
+		for _, f := range pkg.Syntax {
+			tf := pkg.Fset.File(f.Pos())
+			if strings.HasSuffix(tf.Name(), ".go") {
+				files = append(files, f)
+			}
+		}
+
+		// Parse out GoDoc.
+		docPkg, err := doc.NewFromFiles(pkg.Fset, files, pkg.PkgPath)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("doc.NewFromFiles: %v", err)
+		}
+
+		toc = append(toc, &tocItem{
+			UID:  pkg.ID,
+			Name: pkg.PkgPath,
+		})
+
+		pkgItem := &item{
+			UID:     pkg.ID,
+			Name:    pkg.PkgPath,
+			ID:      pkg.Name,
+			Summary: docPkg.Doc,
+			Langs:   onlyGo,
+			Type:    "package",
+		}
+		pkgPage := &page{Items: []*item{pkgItem}}
+		pages[pkg.PkgPath] = pkgPage
+
+		for _, c := range docPkg.Consts {
+			name := strings.Join(c.Names, ", ")
+			id := strings.Join(c.Names, ",")
```

Comment:
Do we need another variable for the same values as name? Looks like name/id are the same for a few of these items.